File: leetcode/p00719/p00719.py
from typing import List

class Solution:
    # Time Complexity: O(nlogn)
    # Space Complexity: O(n)
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        
        def count(distance: int) -> bool:
            acc, cur1, cur2, size = 0, 0, 0, len(nums)
            while cur1 < size:
                while cur2 < size and nums[cur2]-nums[cur1] <= distance:
                    cur2 += 1
                acc += cur2 - cur1 - 1
                cur1 += 1
                if acc >= k:
                    return True
            
            return False
        
        nums.sort()
        low, high = 0, nums[-1] - nums[0]
        while low < high:
            mid = low + (high-low) // 2
            if count(mid):
                high = mid
            else:
                low = mid + 1
        
        return low

if __name__ == '__main__':
    s = Solution()

    assert s.smallestDistancePair([1,3,1], 1) == 0
    assert s.smallestDistancePair([1,1,1], 2) == 0
    assert s.smallestDistancePair([1,6,1], 3) == 5

File: leetcode/p00719/p00719.go
package p00719

import "sort"

// Time Complexity: O(nlogn)
// Space Complexity: O(n)
func smallestDistancePair(nums []int, k int) int {
	sort.Ints(nums)

	low, high := 0, nums[len(nums)-1]-nums[0]
	for low < high {
		mid := low + (high-low)/2
		if count(nums, mid, k) {
			high = mid
		} else {
			low = mid + 1
		}
	}

	return low
}

func count(nums []int, distance, k int) bool {
	acc, left, right, size := 0, 0, 0, len(nums)
	for left < size {
		for right < size && nums[right]-nums[left] <= distance {
			right++
		}
		acc += right - left - 1
		if acc >= k {
			return true
		}

		left++
	}

	return false
}
